src/toast/tod/crosstalk.py

```diff
@@ -0,0 +1,473 @@
+                        if det_lut[name] != rank:
+                            raise RuntimeError(f'Error in creating a LUT from detector name to rank: {det_lut}')
+
+                # mat-mul
+                row_local_total = tod.cache.create(f"{crosstalk_name}_row_local_total_{rank}", np.float64, (n_samples,))
+                if n_local_dets > 0:
+                    row_local_weights = tod.cache.create(f"{crosstalk_name}_row_local_weights_{rank}", np.float64, (n_local_dets,))
+                    local_det_idxs = tod.cache.create(f"{crosstalk_name}_local_det_idxs_{rank}", np.int64, (n_local_dets,))
+                for i, name in enumerate(local_crosstalk_dets_set):
+                    local_det_idxs[i] = names.index(name)
+                # row-loop
+                # * potentially the tod can have more detectors than SimpleCrosstalkMatrix.names_str has
+                # * and they will be skipped
+                for name, row in zip(names, crosstalk_data):
+                    rank_owner = det_lut[name]
+                    if n_local_dets > 0:
+                        row_local_total[:] = 0.
+                        row_local_weights[:] = row[local_det_idxs]
+                        tods_list = [tod.cache.reference(f"{signal_name}_{names[local_det_idxs[i]]}") for i in range(n_local_dets)]
+                        _fma(row_local_total, row_local_weights, *tods_list)
+                    if rank == rank_owner:
+                        row_global_total = tod.cache.create(f"{crosstalk_name}_{name}", np.float64, (n_samples,))
+                        comm.Reduce(row_local_total, row_global_total, root=rank_owner)
+                        # it is reduced into tod.cache and the python reference can be safely deleted
+                        del row_global_total
+                    else:
+                        comm.Reduce(row_local_total, None, root=rank_owner)
+                del det_lut, row_local_total, name, row, rank_owner
+                tod.cache.destroy(f"{crosstalk_name}_row_local_total_{rank}")
+                if n_local_dets > 0:
+                    del row_local_weights, local_det_idxs, tods_list
+                    tod.cache.destroy(f"{crosstalk_name}_row_local_weights_{rank}")
+                    tod.cache.destroy(f"{crosstalk_name}_local_det_idxs_{rank}")
+
+                for name in local_crosstalk_dets_set:
+                    # overwrite it in-place
+                    # not using tod.cache.put as that will destroy and create
+                    tod.cache.reference(f"{signal_name}_{name}")[:] = tod.cache.reference(f"{crosstalk_name}_{name}")
+                    tod.cache.destroy(f"{crosstalk_name}_{name}")
+
+    def exec(
+        self,
+        data: 'toast.dist.Data',
+        signal_name: 'str',
+    ):
+        """Apply crosstalk matrix on ToD in data, dispatched depending if MPI is used."""
+        self._exec_serial(data, signal_name) if IS_SERIAL else self._exec_mpi(data, signal_name)
```

Comment:
The serial version is considerably simpler in execution as it knows it owns everything and not, e.g., constructing the detector LUT. (It should also be (slightly) faster when executed in serial, not that it would matters.)

I actually only implemented the MPI case in the beginning, and later added the serial case to "complete all cases".

I think the serial cases also has a purpose of explaining the algorithm too.

Reply:
I removed the branching in the read class-method. So now the only branch is between `_exec_serial` and `_exec_mpi`. I think this should remains as explained above. If you agree then you can click "Resolve conversation" below.